src/python/pants/backend/python/rules/pex_from_targets.py

```diff
@@ -121,6 +125,43 @@ async def pex_from_targets(request: PexFromTargetsRequest, python_setup: PythonS
         additional_requirements=request.additional_requirements,
     )
 
+    if python_setup.options.resolve_all_constraints:
+        if python_setup.requirement_constraints is None:
+            raise ValueError(
+                "resolve_all_constraints in the [python-setup] scope is set, so "
+                "requirement_constraints in [python-setup] must also be provided."
+            )
+        exact_req_projects = {Requirement.parse(req).project_name for req in exact_reqs}
+        # Add the requirements from the constraint file. Note that this only guarantees
+        # a single resolve if the constraint file covers all the individual requirements
+        # of individual targets. This may be defeated by, e.g., inline requirements.
+        # However we will still do the right thing in that case, it's just the performance
+        # will be impacted by extra resolves.
+        constraint_file_snapshot = await Get(
+            Snapshot,
+            PathGlobs(
+                [python_setup.requirement_constraints],
+                glob_match_error_behavior=GlobMatchErrorBehavior.error,
+                conjunction=GlobExpansionConjunction.all_match,
+                description_of_origin="the option `--python-setup-requirement-constraints`",
+            ),
+        )
+        constraints_file_contents = await Get(
+            DigestContents, Digest, constraint_file_snapshot.digest
+        )
+        constraints_file_reqs = set(
+            parse_requirements(next(iter(constraints_file_contents)).content.decode())
+        )
+        constraint_file_projects = {req.project_name for req in constraints_file_reqs}
```

Comment:
Nit: `FrozenOrderedSet` is preferred because it's deterministic. I don't think it actually matters here because `PexRequirements` will end up sorting the input anyways, but in general, it's a good practice.

Reply:
That seems excessive to me. We purposely ensured order in PexRequirements so that callers wouldn't have to worry about it, no?

Reply:
Fair